python/tvm/runtime/disco/session.py

```diff
@@ -0,0 +1,294 @@
+# under the License.
+"""This module defines a Session in Disco. Session is the primary interface that users interact
+with the distributed runtime.
+"""
+from typing import Any, Callable, Optional, Sequence
+
+from ..._ffi import register_object
+from ..._ffi.runtime_ctypes import Device
+from ..ndarray import NDArray
+from ..object import Object
+from . import _ffi_api
+
+
+@register_object("runtime.disco.DRef")
+class DRef(Object):
+    """An object that exists on all workers. The controller process assigns a unique "register id"
+    to each object, and the worker process uses this id to refer to the object residing on itself.
+    """
+
+    @property
+    def session(self) -> "Session":
+        """Get the session that this DRef belongs to."""
+        return _ffi_api.DRefSession(self)  # type: ignore # pylint: disable=no-member
+
+    def debug_get_from_remote(self, worker_id: int) -> Any:
+        """Get the value of a DRef from a remote worker.
+
+        Parameters
+        ----------
+        worker_id : int
+            The id of the worker to be fetched from.
+
+        Returns
+        -------
+        value : object
+            The value of the register.
+        """
+        return _ffi_api.DRefDebugGetFromRemote(self, worker_id)  # type: ignore # pylint: disable=no-member
+
+
+class DPackedFunc(DRef):
+    """A PackedFunc in a Disco session."""
+
+    def __init__(self, dref: DRef) -> None:
+        self.handle = dref.handle
+        dref.handle = None
+
+    def __call__(self, *args) -> DRef:
+        return self.session.call_packed(self, *args)
+
+
+class DModule(DRef):
+    """A Module in a Disco session."""
+
+    def __init__(self, dref: DRef) -> None:
+        self.handle = dref.handle
+        del dref.handle
+
+    def __getitem__(self, name: str) -> DPackedFunc:
+        func = self.session._get_cached_method("runtime.ModuleGetFunction")
+        return DPackedFunc(func(self, name, False))
+
+
+@register_object("runtime.disco.Session")
+class Session(Object):
+    """A Disco interactive session. It allows users to interact with the Disco command queue with
+    various PackedFunc calling convention."""
+
+    @staticmethod
+    def threaded_session(num_workers: int) -> "Session":
+        """Create a threaded session."""
+        return _ffi_api.SessionThreaded(num_workers)  # type: ignore # pylint: disable=no-member
+
+    def _get_cached_method(self, name: str) -> Callable:
+        if not hasattr(self, "_cache"):
+            cache = self._cache = {}  # pylint: disable=attribute-defined-outside-init
+        else:
+            cache = self._cache
+        if name not in cache:
+            func = cache[name] = self.get_global_func(name)
+        else:
+            func = cache[name]
+        return func
+
+    def empty(
+        self,
+        shape: Sequence[int],
+        dtype: str,
+        device: Optional[Device] = None,
+    ) -> DRef:
+        """Create an empty NDArray on all workers.
```

Comment:
The implementation of runtime.disco.empty does not match the document. It only creates an empty array on controller instead of workers.

Reply:
This is indeed creating an array on the worker side, and attach a DRef that references these array. 

Perhaps we can update the docs to

```
Create an empty NDArray on all workers and attach them to a  DRef
```

Reply:
ok I was wrong.